### ISSUE TITLE
Add support for pandas and polars dataframes

### ISSUE THREAD

**Issue by Galedon1** [pyvista/pyvista#5516]
### Describe the feature you would like to be added.

Polars and especially Pandas are very core Python data processing libraries by now. It would be nice to be able to create a mesh and/or add scalars directly from these datasets instead of the (very costly) casting to Numpy arrays.
Now the Polydata creation raises: 
```
TypeError: Invalid Input type:

Expected first argument to be either a:
- vtk.PolyData
- pyvista.PolyData
- numeric numpy.ndarray (1 or 2 dimensions)
- List (flat or nested with 3 points per vertex)
- vtk.vtkDataArray

Instead got: <class 'polars.series.series.Series'>
```

And the scalar assigning takes ridiculously long time (did not manage to see if it works in the end)

### Pseudocode or Screenshots

```python
import pyvista as pv
import polars as pl
import numpy as np

N = int(1e6)
data = pl.DataFrame(dict(
    points = np.random.sample((N, 3)),
    color = np.random.sample((N, 4))
    ))

cloud = pv.PolyData(data["points"])
cloud["colors"] = data["color"]
```

**Comment by darikg** [comment]
```python
data["points"].to_numpy(zero_copy_only=True)
```
`ValueError: cannot return a zero-copy array`

I think polars is storing the data as a heterogenous list-of-lists so there's no way to get a reference to a single contiguous buffer. Forcing polars to use array storage seems to work:

```python
data = pl.DataFrame(dict(
        points=orig_pts,
        color=np.random.sample((N, 4))
    ),
    schema=dict(
        points=pl.Array(pl.Float64, 3),
        color=pl.Array(pl.Float64, 4),
    ),
)

mesh = pv.PolyData(data['points'].to_numpy(zero_copy_only=True))
mesh['color'] = data['color'].to_numpy(zero_copy_only=True)
```
I *think* in this case polars defaults to zero-copy anyway. I think to support pandas and polars more fluently all that needs to be done is removing explicit `isinstance(arr, ndarray)` with just letting `numpy.asarray` do it's thing, since polars and pandas support the numpy array interface

**Comment by tkoyama010** [comment]
I agree about the fact that the process remains same. The good thing about this request is that it simplifies the API calls for the user in my mind. Supporting many types can be burdensome. However, pandas and polar are so common that it would be beneficial to add them.

**Comment by banesullivan** [comment]
Reminds me of #265

**Comment by user27182** [comment]
I am not familiar with pandas or polars data structures. However, it sounds like it may be possible to include these data types as part of a more general `ArrayLike` definition.

Currently, `PyVista`'s `ArrayLike` definition is `np.ndarray` and `Sequence` centric:
https://github.com/pyvista/pyvista/blob/09cc31d9d256b33067a89cec5060ebf4eef8d90f/pyvista/core/_typing_core/_array_like.py#L41-L45

But perhaps this definition can be broadened to accommodate other array-like types directly.

I think the main issue would be pre-processing any array-like inputs in a suitable and consistent manner to ensure minimal casting is done to `NumPy`, or at least done in manner with `zero_copy_only` or simply doing `np.asarray` as suggested above.

I'll note that an input validation package for `PyVista` is currently in active development, and in #5226 an array-like parsing function is added [here](https://github.com/pyvista/pyvista/blob/eac7c96f3f0606e769f21109c834c66f1bc52387/pyvista/core/validation/_array_like.py#L121-L254).
This parsing function aims to extract minimal info about the array-like input without doing any casting (while also ensuring the output type can be statically inferred by Mypy). Again, perhaps this could be extended to accept `pandas`or `polars` types and do minimal pre-processing so that the output of something like `validate_array(array: ArrayLike)` has a standard `Sequence` or `np.ndarray`  representation.

**Comment by darikg** [comment]
I have a proof-of-concept [here](https://github.com/pyvista/pyvista/compare/pyvista:pyvista:main...darikg:pyvista:maint/cell-array-tests), it just removes the explicit checks for isinstance(ndarray) and passes the inputs into `validate_array*`, so the conversion is done entirely by np.asarray. Using a mock array class like this

```python
class ArrayLikeWrapper:
    """A class that implements the NumPy array protocol but isn't isinstance(np.ndarray)"""
    def __init__(self, arr):
        self._arr = np.asarray(arr)

    def __getattr__(self, item):
        return getattr(self.__getattribute__('_arr'), item)
```

everything "just works", which I believe should work for pandas and polars data. @user27182 I *think* we would just need to add numpy.typing._SupportsArray[_NumType] to our _ArrayLike definition.

**Comment by user27182** [comment]
I thought `np.asarray` _always_ copies the input unless the input is type `np.ndarray`?? Is that not the case? For pure `Sequence` inputs, copies would definitely be made, right?

I agree a wrapper of some sort is probably good to use for arraylike inputs... in #5226 a `namedtuple` container is currently used for that purpose.

And if we're using protocols, then doing `isinstance(array, _SupportsArray)` should work directly, no? If that's the case then perhaps a wrapper class isn't necessary, just swap out `np.ndarray` with `_SupportsArray` in the instance checks.  The issue goes back to supporting `Sequence` types natively though without copying/casting to `np.ndarray`...

Part of what I'm trying to with with #5226 is remove the existing 'cast everything to a numpy array first` behaviour of the input validation package. This is so that we can support something like:

``` python
ArrayLikeType = TypeVar('ArrayLikeType', bound=ArrayLike)
def validate_array(array: ArrayLikeType) -> ArrayLikeType: ...
```

Where the output array type is always the same as the input array type. This is so that doing `validate_array`, say, on a 3-Tuple point array, would do any or all checks as needed and return the exact same 3-tuple input as the output. The output would only be cast to a specific array-like type, e.g. cast to `np.ndarray`, if explicitly set as a kwarg option. Ideally, if supporting other array-like types, the same no-copy behaviour would apply.

**Comment by darikg** [comment]
> I thought `np.asarray` _always_ copies the input unless the input is type `np.ndarray`?? Is that not the case? For pure `Sequence` inputs, copies would definitely be made, right?

I think that's up to the individual implementations of the array protocol? e.g. for pandas

```python
import pandas as pd
import numpy as np
x = pd.Series([1, 2, 3])
y = np.asarray(x)
y[0] = 999
assert x[0] == 999
```

passes. Trying the same thing on a polars.Series you get `ValueError: assignment destination is read-only` -- I assume polars is explicitly setting the array view read-only to prevent mutating the original array.

**Comment by user27182** [comment]
@darikg Thanks for the example code.

I believe this works with `Pandas` because it seems like the `Series` object stores its values as a numpy array internally. So, calling `np.asarray` does not make a copy since the values `[1,2,3]` were _already_ copied when the object was created.

The internal series array representation seems to be stored as a property named `_values`. See the pandas `__array__` code [here](https://github.com/pandas-dev/pandas/blob/f538741432edf55c6b9fb5d0d496d2dd1d7c2457/pandas/core/series.py#L973-L1025).

We can verify the series values are store as a numpy array with:
``` python
import pandas as pd
import numpy as np
x = pd.Series([1, 2, 3])
assert type(x._values) == np.ndarray  # True
```

For `polars`, they have a similar `__array__` definition [here](https://github.com/pola-rs/polars/blob/801da67446d6151e197595e1fa4bd354bd2606fe/py-polars/polars/dataframe/frame.py#L1317-L1327) for `DataFrame`.
Unfortunately though, unlike `pandas`, it seems that `polars` does not store its data internally as a numpy array. Therefore, it's necessary to use the  `zero_copy_only` kwarg with `to_numpy()` directly to avoid copying the data. Simply using `np.asarray` on a polars Series or DataFrame, which will invoke `__array__`, will call `to_numpy()` without any kwargs. And by [default](https://github.com/pola-rs/polars/blob/801da67446d6151e197595e1fa4bd354bd2606fe/py-polars/polars/series/series.py#L4289-L4295), `zero_copy_only` is False.  So I believe a copy will always be made this way. E.g.:
``` python
import numpy as np
import polars as pl
x = pl.Series("a", [1, 2, 3])
y = np.asarray(x)
assert id(x[0]) == id(y[0])  # AssertionError
```

So, it definitely appears as though we _should_ be able to support pandas and polars series / dataframes directly using the `_SupportsArray` Protocol, and by simply using `np.asarray` directly without much else.

``` python
from typing import runtime_checkable, Protocol, Any, TypeVar
import pandas as pd
import numpy as np
import polars as pl

_DType_co = TypeVar('_DType_co', bound=np.generic, covariant=True)

@runtime_checkable
class _SupportsArray(Protocol[_DType_co]):
    def __array__(self) -> np.ndarray[Any, _DType_co]: ...

x = pd.Series([1, 2, 3])
y = pl.Series('a', [1, 2, 3])
assert isinstance(x, _SupportsArray)  # True
assert isinstance(y, _SupportsArray)  # True
```

However, using this Protocol abstraction will result in the data being copied for `polars` data type inputs when using `np.asarray`. To use avoid copying for `polars` data I believe an explicit use of `to_numpy(zero_copy_only=True)` may be required.

**Comment by darikg** [comment]
hmm I think something's weird with that id assertion

```python
import numpy as np
import polars as pl
x = pl.Series("a", [1, 2, 3])
y = x.to_numpy(zero_copy_only=True)
assert id(x[0]) == id(y[0])  # AssertionError
```

I think it's checking the ids of the values after the `__getitem__`, which isn't nessecarily the same thing as the actual array element

my reading of polars.Series.to_numpy there is that it always attempts a zero-copy conversion regardless of zero_copy_only, but without zero_copy_only=True it will silently copy if *necessary* (instead of raising an error).

**Comment by user27182** [comment]
Good point, yes in the docs it states "data which is purely numeric AND without null values is not cloned"

**Comment by user27182** [comment]
`polars` sets the `writeable` flag to False by default with its `to_numpy` method. This can be manually changed. If we do this, we can see that `np.asarray` does not copy the `polars` object data.
``` python
import numpy as np
import polars as pl
x = pl.Series("a", [1, 2, 3])
y = np.asarray(x)
y.flags['WRITEABLE']=True
y[0] = 999
assert x[0] == 999
```

So, in line with @darikg's earlier suggestions, supporting `polars` or `pandas` data types in `pyvista` could look something like:

1. Use `np.asarray` on all array-like inputs 
(Or, use `isinstance(array, _SupportsArray)` first if specific logic is needed, e.g. for handling `Sequence` types separately which we may want to do to avoid creating an unnecessary copy)
2. Add `numpy.typing._SupportsArray[_NumType]` to our `_ArrayLike` definition 
(Or, directly replace the `NumpyArray` alias with `_SupportsArray` instead)

I don't think (2) is something that should necessarily be added at the moment, though. This is because in my view calling `np.asarray` is strictly necessary to support these types. If we change the `ArrayLike` annotations, then any functions which use them must be explicitly aware of the fact that calling `np.asarray` is needed to ensure `_SupportsArray` inputs will work as expected with any numpy array operations downstream.  Even if we changed any/all functions which currently use `ArrayLike` annotations to use `np.asarray` explicitly, how will this requirement be enforced for functions which add `ArrayLike` annotations in the future?

Another issue is that calling `np.asarray` can create unexpected errors, e.g. calling `np.asarray` directly on the input given in the original post:

``` python
import polars as pl
import numpy as np

N = int(1e6)
data = pl.DataFrame(dict(
    points = np.random.sample((N, 3)),
    color = np.random.sample((N, 4))
    ))
array = np.asarray(data["points"])
```

will yield:
```polars.exceptions.ComputeError: 'to_numpy' not supported for dtype: List(Float64)```

Which I think is a bit puzzling as to what the cause may be or what the possible workaround may be. Should we try...except for this to provide a better message? If so, now we would probably need to wrap the try..except block inside a custom pyvista `asarray` function...

An alternative solution is to support these types, but only as part of pyvista's input validation functions. There, the need to use `asarray` or do `try..except` can be localized.

**Comment by user27182** [comment]
I noticed that `polars` has a bunch of Export APIs, e.g. for [Series](https://docs.pola.rs/py-polars/html/reference/series/export.html) and for [DataFrame](https://docs.pola.rs/py-polars/html/reference/dataframe/export.html).

Perhaps the requested feature could be implemented directly in `polars` instead of `pyvista`? E.g. by adding a `to_pyvista` export method. @Galedon1 perhaps it's worth making a feature request on the polars github for this?

Another option is to add a dedicated function in `pyvista` for this, e.g. add `polars_to_pyvista` or `pandas_to_pyvista` in `pyvista.core.utilities.arrays`.

**Comment by Galedon1** [comment]
> Perhaps the requested feature could be implemented directly in `polars` instead of `pyvista`? E.g. by adding a `to_pyvista` export method. @Galedon1 perhaps it's worth making a feature request on the polars github for this?

I considered it, but wouldn't it just became an alias for `to_numpy`?

**Comment by user27182** [comment]
> I considered it, but wouldn't it just became an alias for `to_numpy`?

@Galedon1 Yea you're right, I suppose we don't need a `to_pyvista` method. I guess the reason I suggested this stems from the fact that trying to use `np.asarray` directly will raise an error.  Using a schema will resolve this as @darikg showed, but this does not seem like an obvious resolution to me. So, perhaps a `to_pyvista` function would be used to handle this kind of thing more gracefully.

**Comment by user27182** [comment]
> Using a mock array class like this
> 
> ```python
> class ArrayLikeWrapper:
>     """A class that implements the NumPy array protocol but isn't isinstance(np.ndarray)"""
>     def __init__(self, arr):
>         self._arr = np.asarray(arr)
> 
>     def __getattr__(self, item):
>         return getattr(self.__getattribute__('_arr'), item)
> ```

I've expanded on @darikg's idea above as part of #5226 and added an array-like wrapper class.
 It still does instance checks so that sequences and scalars can be handled separately, but it also has a 'catch-all' case where any input that isn't a common case will get cast as a numpy array. This means that instances of the `_SupportsArray` protocol are implicitly accepted (including `pandas` and `polars` dataframes). Here's the key bits of code:
https://github.com/pyvista/pyvista/blob/2728c74dcdee3f6c38603364e4a3ef1f59b9d576/pyvista/core/validation/_array_wrapper.py#L142-L152

Once that PR gets merged we could make use of the validation functions in key areas like `pv.DataSet` constructors or `pv.plot` to better support arbitrary `_SupportsArray` data.

**Comment by user27182** [comment]
`sklearn` has a [check_array](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html) function which has a lot of pandas-related checks in the source code to format the output. Maybe just using that directly before passing data to pyvista is good enough?

**Comment by darikg** [comment]
> I've expanded on @darikg's idea above as part of #5226 and added an array-like wrapper class.

Just to clarify -- I was only suggesting the use of a mock wrapping for testing to ensure that anything array-like worked, without having the tests need a pandas or polars dependency, etc.

For passing array-likes for mesh points, cell array, or point/cell data, ultimately everything needs to be converted to a `vtkDataArray`. I think currently pyvista almost exclusively relies on `vtkmodules.util.numpy_support.numpy_to_vtk`, so `np.asarray` seems like a more or less necessary stepping stone. If a zero-copy is available, this is pretty minimal overhead, and if it's not available because the data is stored non-contiguously, (as in the `polars.List[polars.List[float]]`) case) I don't think any hand-crafted conversion is going to be much better than copying to numpy and converting, because the data needs to be copied to a contigous buffer no matter what.

**Comment by user27182** [comment]
I found that calling `to_list` seems to work to cast the data to a list format. The list can then be cast as a numpy array without issue. This creates a copy, but at least it will work without error.

PR #5567 adds this workaround and prints a warning about the data being copied. It also attempts to gracefully handle any other errors which may be generated by `polars`.
